### ISSUE TITLE
Parallel reducer, `MDRangePolicy` and HIP or CUDA: doesn't work with bool, char, int16_t

### ISSUE THREAD

**Issue by romintomasetti** [kokkos/kokkos#5641]
**Describe the bug**

It seems that `Kokkos::LAnd<bool>` does not work with `Kokkos::Experimental::HIP` and `Kokkos::MDRangePolicy`.

**Please also include the following items to support reproducing the bug**

* Kokkos release 3.7
* Linux with ROCm 5.3 (vega906)
* code needed to reproduce the bug
   ```c++
   template <typename T, typename execution_space, typename policy_t, typename functor_t>
   T parallel_reducer_LAnd(const policy_t& policy, const functor_t& f)
   {
      T allEqual = 0;
      auto reducer = Kokkos::LAnd<T>(allEqual);
      Kokkos::parallel_reduce(policy,f,reducer);
      execution_space().fence();
      return allEqual;
   }

   TEUCHOS_UNIT_TEST(test,test)
   {
      std::cout << "KOKKOS_VERSION: " << KOKKOS_VERSION << std::endl;

      using execution_space = Kokkos::Experimental::HIP;

      const auto  policy_MD = Kokkos::MDRangePolicy<Kokkos::Rank<2>,execution_space>({0,0},{8,2});
      const auto  policy_RG = Kokkos::  RangePolicy<                execution_space>( 0,    8   );

      const auto res_MD_bool = parallel_reducer_LAnd<bool,execution_space>(policy_MD, KOKKOS_LAMBDA(const size_t, const size_t, bool& res){res = true;});
      const auto res_MD_int  = parallel_reducer_LAnd<int ,execution_space>(policy_MD, KOKKOS_LAMBDA(const size_t, const size_t, int & res){res = 1   ;});

      const auto res_RG_bool = parallel_reducer_LAnd<bool,execution_space>(policy_RG, KOKKOS_LAMBDA(const size_t, bool& res){res = true;});
      const auto res_RG_int  = parallel_reducer_LAnd<int ,execution_space>(policy_RG, KOKKOS_LAMBDA(const size_t, int & res){res = 1   ;});

      TEST_EQUALITY(res_MD_bool,false); // -> I expected true here
      TEST_EQUALITY(res_RG_bool,true);
      TEST_EQUALITY(res_MD_int ,1);
      TEST_EQUALITY(res_RG_int ,1);
   }
   ```
7. please also attach the `KokkosCore_config.h` header file (generated during the build); [KokkosCore_config.hpp.txt](https://github.com/kokkos/kokkos/files/10102708/KokkosCore_config.hpp.txt)

As you can see from the reproducer, using the `Kokkos::LAnd<bool>` with `MDRangePolicy` on HIP seems not to work correctly.

A temporary workaround is to use `int` instead.

Please note that for `OpenMP` and `Cuda` it works fine.

**Comment by ajpowelsnl** [comment]
@Rombur - have you had a chance yet to look into this issue?

**Comment by Rombur** [comment]
I am working on it.

**Comment by PhilMiller** [comment]
Should this be treated as release-critical? Silent wrong results are very bad for users. At the very least, if we can't immediately fix this, we should add a `static_assert` to fail the build in the dangerous case, suggesting the workaround.

**Comment by PhilMiller** [comment]
Comment typo noticed while investigating a guardrail against this:
https://github.com/kokkos/kokkos/blob/master/core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp#L280
'arithmatically'

**Comment by PhilMiller** [comment]
Proposed guardrail in https://github.com/kokkos/kokkos/pull/5745

**Comment by PhilMiller** [comment]
```c++
      int bound = 2;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
      {
      const auto  policy_MD = Kokkos::MDRangePolicy<Kokkos::Rank<2>,execution_space>({0,0},{bound,2});
      const auto  policy_RG = Kokkos::  RangePolicy<                execution_space>( 0,    bound   );

      int k = 0;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               
        {
        const auto lor_MD_int16  = parallel_reducer_LOr<uint16_t ,execution_space>(policy_MD, KOKKOS_LAMBDA(const size_t i, const size_t j, uint16_t & res){res = i == k;});
        EXPECT_EQ(lor_MD_int16 ,1) << bound << " " << k << "\n";
      }
    }
  }
```

That fails, with any of `char`, `signed char`, `unsigned char`, `(u)int{8,16}_t` in place of `uint16_t`.

**Comment by PhilMiller** [comment]
It passes with `int32_t` or `uint32_t`.

**Comment by PhilMiller** [comment]
Maybe see discussion in https://github.com/kokkos/kokkos/pull/5324 for a lead on a fix?

**Comment by PhilMiller** [comment]
Have a look at #5333/#5334 for thoughts on a possible cause/fix

**Comment by PhilMiller** [comment]
I'm going to open a separate issue for the OpenMPTarget backend, since its implementation is distinct, and @rgayatri23 asked me to just disable to test in question for now.

**Comment by masterleinad** [comment]
Fixed in #5641.

**Comment by dalg24** [comment]
I suppose you mean #5745
